arrows/vxl/aligned_edge_detection.cxx

```diff
@@ -0,0 +1,315 @@
+// This file is part of KWIVER, and is distributed under the
+// OSI-approved BSD 3-Clause License. See top-level LICENSE file or
+// https://github.com/Kitware/kwiver/blob/master/LICENSE for details.
+
+#include "aligned_edge_detection.h"
+
+#include <arrows/vxl/image_container.h>
+
+#include <vil/algo/vil_gauss_filter.h>
+#include <vil/algo/vil_sobel_3x3.h>
+#include <vil/vil_convert.h>
+#include <vil/vil_image_view.h>
+#include <vil/vil_math.h>
+#include <vil/vil_plane.h>
+
+#include <limits>
+#include <type_traits>
+
+#include <cstdlib>
+
+namespace kwiver {
+
+namespace arrows {
+
+namespace vxl {
+
+// ----------------------------------------------------------------------------
+/// Private implementation class
+class aligned_edge_detection::priv
+{
+public:
+  priv( aligned_edge_detection* parent ) : p{ parent } {}
+
+  template < typename PixType, typename GradientType >
+  vil_image_view< PixType >
+  calculate_aligned_edges( vil_image_view< PixType > const& input,
+                           vil_image_view< GradientType >& grad_i,
+                           vil_image_view< GradientType >& grad_j );
+
+  template < typename InputType, typename OutputType >
+  vil_image_view< OutputType >
+  nonmax_suppression( vil_image_view< InputType > const& grad_i,
+                      vil_image_view< InputType > const& grad_j );
+
+  template < typename pix_t > vil_image_view< pix_t >
+  filter( vil_image_view< pix_t > const& input_image );
+
+  aligned_edge_detection* p;
+  // Internal parameters/settings
+  float threshold = 10;
+  bool produce_joint_output = true;
+  double smoothing_sigma = 1.3;
+  unsigned smoothing_half_step = 2;
+};
+
+// ----------------------------------------------------------------------------
+// Perform NMS on the input gradient images in horizontal and vert directions
+// only
+template < typename InputType, typename OutputType >
+vil_image_view< OutputType >
+aligned_edge_detection::priv
+::nonmax_suppression( vil_image_view< InputType > const& grad_i,
+                      vil_image_view< InputType > const& grad_j )
+{
+  if( grad_i.ni() != grad_j.ni() || grad_i.nj() != grad_j.nj() )
+  {
+    LOG_ERROR(
+      p->logger(),
+      "Input gradient image dimensions must be equivalent" );
+  }
+
+  auto const ni = grad_i.ni();
+  auto const nj = grad_i.nj();
+
+  vil_image_view< OutputType > output( ni, nj, 2 );
+  output.fill( 0 );
+
+  // Perform non-maximum suppression
+  for( decltype(+nj) j = 1; j < nj - 1; j++ )
+  {
+    for( decltype(+ni) i = 1; i < ni - 1; i++ )
```

Comment:
```suggestion
  for( decltype(+nj) j = 1; j < nj - 1; ++j )
  {
    for( decltype(+ni) i = 1; i < ni - 1; ++i )
```

(Aside: I usually try to write these using `range::iota`, but there isn't a version that starts at non-zero.)